include/revng/ABI/Definition.h

```diff
@@ -274,6 +274,10 @@ class Definition : public generated::Definition {
 public:
   using generated::Definition::Definition;
 
+private:
+  friend struct KeyedObjectTraits<abi::Definition>;
```

Comment:
Why is this necessary? The generated fields are public.

Reply:
Whoops, it seems I have missed this one on my first go-through. Sorry.

It's specifically for the constructor. The generated `KeyedObjectTraits<abi::Definition>` wants to be able to access a way to construct the object from it's key.

But, in this case, the key is just a `model::ABI::Values`, so such a construction results in an empty `abi::Definition` object which led to some pretty hard to debug problems 😅.

TL;DR: we can't let `abi::Definition` to be constructible from key alone. Or be constructible at all, it should always come from the `abi::Definition::get` thing, but we still have to make `KeyedObjectTraits` happy. This is the least invasive solution I could come up with. Do you have any better ideas?

Reply:
The idea of TTG data structures is to be default constructible and constructible from the key and have all public fields.
They are basically plain old `struct`s.

Any other concerns about the *validity* of the content, should be verified through a `verify` method.
Please, let's not deviate from this paradigm. It makes refactoring of TTGs hard, coherence over minor improvements.

Reply:
Then I guess TTG is not a good fit for the definition, is it?

Consider the case
```cpp
void some_function_that_wants_a_definition(const abi::Definition &ABI);

// Some clueless user just passes the thing they have named `ABI` in:
void clueless_user() {
  model::ABI::Values /* std::string eventually */ ABI = getABIFromSomewhere();
  some_function_that_wants_a_definition(ABI);
}
```
And because the definition is just a "plain old" struct, this leads to a lot of sadness.

Consider that verification does not really cut it here either since there's no way to tell a legit really basic definition with barely any fields filled from a newly created empty one. That would require a really awful, long and prone to getting outdated verification routine that would literally be equivalent to "try and find any field that is not default initialized".

Reply:
I'm fine with that situation, if a very simple empty ABI is valid, then so be it.

You can put a verify that checks that we need at least an argument register.
It's not like your observation is not appropriate, but deviating from convention when around TTGs has already made my life harder in the past (specifically: reusing `efa::` stuff in `yield::` stuff), so let's not do that again.

TTGs are great for certain things and suboptimal for others, we can have paths towards improvement, but these deviation from simple usages are obstacles along this path.

Reply:
> we need at least an argument register

x86 SystemV ABI already fails.

Reply:
You can't make up a rule that ensures a default constructed object is not valid?

Even if we can't, what said above holds, let's keep it simple and make it publicly default constructible and constructible from a key.

Reply:
> You can't make up a rule that ensures a default constructed object is not valid?

I guess there's `StackAlignment` field that should not be set to 0. But that sounds finicky at best.

EDIT: wait, no, if we ever want to support linux syscall ABI, for example, it doesn't allow any arguments in stack - so it would make sense for the `StackAlignment` in that ABI to be set to 0.

The main problem is that I can *think* of an ~~fictional~~ ABI that doesn't need anything.
For example, what if it only ever passes arguments by placing them on a predefined address somewhere in memory. Sure it probably doesn't exist, but if it did, it wouldn't need to specify anything - since we have no way to recover such argument anyway - all the functions would be `void function(void)` from our point of view.

Reply:
Oh well, still, let's stick to the convention please.
I don't think anyone will be actually affected by this.

Reply:
Dropped it :sadparrot: